rcode-reader/src/main/java/eu/rssw/pct/PLReader.java

```diff
@@ -0,0 +1,198 @@
+
+    private File pl;
+    private List<FileEntry> files = null;
+    private boolean mm = false;
+
+    public PLReader(File file) {
+        this.pl = file;
+    }
+
+    /**
+     * Returns entries contained in this procedure library
+     * 
+     * @throws RuntimeException If file is not a valid procedure library
+     */
+    public List<FileEntry> getFileList() {
+        if (this.files == null)
+            readFileList();
+        return files;
+    }
+
+    public FileEntry getEntry(String name) {
+        for (FileEntry entry : getFileList()) {
+            if (entry.getFileName().equals(name))
+                return entry;
+        }
+
+        return null;
+    }
+
+    private void readFileList() {
+        try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+            FileChannel fc = raf.getChannel();
+            int version = 0;
+            ByteBuffer magic = ByteBuffer.allocate(2);
+            fc.read(magic);
+            int magicVal = magic.getShort(0) & 0xffff;
+            if (magicVal == MAGIC) {
+                version = 1;
+            } else if (magicVal == MAGIC_V11) {
+                version = 2;
+            } else if (magicVal == MAGIC_V11_MM) {
+                version = 2;
+                mm = true;
+            } else {
+                throw new InvalidLibraryException("Incorrect magic number");
+            }
+
+            Charset charset = getCharset(fc);
+            int offset = getTOCOffset(fc, version);
+            files = new ArrayList<>();
+            FileEntry fe = null;
+            while ((fe = readEntry(fc, offset, charset, version)) != null) {
+                if (fe.isValid())
+                    files.add(fe);
+                offset += fe.getTocSize();
+            }
+        } catch (IOException caught) {
+            throw new InvalidLibraryException(caught);
+        }
+    }
+
+    public boolean isMemoryMapped() {
+        return mm;
+    }
+
+    public InputStream getInputStream(FileEntry fe) throws IOException {
+        ByteBuffer bb = null;
+        try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+            FileChannel fc = raf.getChannel();
+            bb = ByteBuffer.allocate(fe.getSize());
+            fc.read(bb, fe.getOffset());
+        }
+
+        return new ByteArrayInputStream(bb.array());
+    }
+
+    private Charset getCharset(FileChannel fc) throws IOException {
+        ByteBuffer bEncoding = ByteBuffer.allocate(ENCODING_SIZE);
+        if (fc.read(bEncoding, ENCODING_OFFSET) != ENCODING_SIZE)
+            throw new InvalidLibraryException("Invalid charset");
+        bEncoding.position(0);
+        StringBuilder sbEncoding = new StringBuilder();
+        int zz = 0;
+        while ((zz < 20) && (bEncoding.get(zz) != 0)) {
+            sbEncoding.append((char) bEncoding.get(zz++));
+        }
+        try {
+            return Charset.forName(sbEncoding.toString());
+        } catch (IllegalArgumentException iae) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1166)